### ISSUE TITLE
Lua API Performance Improvement

### ISSUE THREAD

**Issue by paradust7** [minetest/minetest#12437]
I got an extremely surprising result from profiling on Linux X86-64 with LuaJIT, running MineClone2 in single-player mode.

In this setup, EmergeThread spends about 53% of its time in ScriptApiEnv::environment_OnGenerated. This is the method that runs Lua callbacks after mapgen, in this case, specific to MineClone2.

However, of that time, 98% of it is spent in C++ land! Only 2% is spent running actual Lua.

Most of the time is being lost to allocation of tables and strings, field look ups, and various other Lua API calls that are happening in the core script api functions. Methods such as: lua_createtable, lua_pushstring, lua_getfield, lua_setfield, ...

Inside of Lua, the JIT is likely able to effectively optimize out tables and table lookups when table keys are known at compile-time. (For example, with a table like {x=x, y=y, z=z}). Returning a string from a table lookup inside of Lua is also cheap, as they are immutable and gc'd. But when coming from C++, the Lua API must do allocations and run-time lookups for everything.

This suggests a way to speed things up:
1) Move the front-end API into Lua, and make a back-end API that only involves passing simple data, like integers, floats, and strings. Then no table or array construction needs to happen from C++.
2) When a core API function returns a string, return a string that is cached in Lua, when possible. Minimize how often strings are passed from C++ to Lua.

I have an initial PR of this change applied to just the single hottest function, core.get_node(). This one change alone takes about 6% off the time needed to run emerge!

**Comment by TurkeyMcMac** [comment]
Do you know how much of a penalty the C-Lua boundary itself is? Perhaps the speed of `pushnode` would increase if C++ were to delegate to a Lua snippet to create the table. This would save a lot of refactoring work.

**Comment by sfan5** [comment]
Did you check how often `get_node` is called? For bulk operations vmanips already exist and I have a few doubts that MCL2 is written efficiently in that regard.

> When a core API function returns a string, return a string that is cached in Lua, when possible. Minimize how often strings are passed from C++ to Lua.

I thought Lua already had string interning? 
Performance of pure Lua code can't be reached of course but worth checking what exactly inside LuaJIT consumes time. If strings are indeed an issue maybe we can do something to reuse common strings.

> However, of that time, 98% of it is spent in C++ land! Only 2% is spent running actual Lua.

Are you sure JITed code appears correctly in the backtraces/flamegraph/whatever you're using to look at it?
I can find 4.4% of "?? [??]" which is what you'd expect for symbol less code but it's not grouped correctly under lua_pcall (or anywhere) but at the very bottom next to `__GI___clone`.

**Comment by sfan5** [comment]
This is my profiling test setup btw #12441

**Comment by paradust7** [comment]
@TurkeyMcMac Moving between C and Lua code is fast, but actual table/array creation and key lookups are slow. Calling a Lua helper from pushnode() to create the table would actually not be any faster... the goal is to keep the table creation and return entirely in Lua so that it can be entirely optimized out by the JIT. When a table crosses the C/Lua boundary, it can't be optimized out.

**Comment by paradust7** [comment]
@sfan5 I use `perf record --call-graph dwarf` combined with `-g -ggdb -fno-omit-frame-pointer`, which means that each sample taken by the kernel includes a full stack trace. This makes it possible to generate a hierarchy:

https://gist.github.com/paradust7/2e6d011c810431b8b2fbe74a02bc93ac

The way to read this is that 51.57% of samples taken had EmergeThread::run in the stack trace. Since samples are taken across all threads, that's the ratio of CPU spent in the emerge thread vs. other threads in the process.

Meanwhile, 28.55% of samples had EmergeGen::finishGen in the stack trace. This means 28.55/51.57 = 55% of the time in EmergeThread was spent in finishGen.

If you look at the relationship between `environment_OnGenerated`, `lua_pcall`, "0x7f5225739016" (the jit'd code), and `script_exception_wrapper`, you can see that 100% of the samples measured from within lua_pcall code also had script_exception_wrapper in the stack. Another way of seeing this, is to pull out just the top-level functions:
```
                  - 27.84% lua_pcall
                     - 27.38% 0x7f1c27acf016
                        - script_exception_wrapper
                           - 5.43% ModApiEnvMod::l_get_node
                           - 4.68% ModApiEnvMod::l_find_nodes_in_area
                           - 3.52% LuaVoxelManip::l_get_param2_data
                           - 2.67% LuaVoxelManip::l_set_param2_data
                           - 2.62% LuaVoxelManip::l_set_data
                           - 2.49% LuaVoxelManip::l_get_data
                           - 1.94% LuaVoxelManip::l_calc_lighting
                           - 1.07% LuaVoxelManip::l_update_liquids
                           - 1.04% ModApiEnvMod::l_find_nodes_in_area_under_air
                           - 0.75% ModApiEnvMod::l_set_node
```
If you sum up the % for those 10 methods, you get 26.21% of samples total. Compared to 27.84% spent in lua_pcall total. So that's 94% of time spent in the ModApiEnvMod/LuaVoxelManip methods. (this is a different run, the percentages bounce around a bit).

**Comment by paradust7** [comment]
@sfan5

I did not measure the # of times get_node is called, I assume it is called for every node in the chunk. So very often.

But the point is, this simple change makes core.get_node() about 8x faster, without changing the interface or how it is used. I have not looked at the rest of the API, but it is likely everything will benefit from applying the same principles. (certainly everything that uses pushnode).

**Comment by sfan5** [comment]
> I use `perf record --call-graph dwarf` combined with `-g -ggdb -fno-omit-frame-pointer`, which means that each sample taken by the kernel includes a full stack trace.

I see, that makes sense. I only use the default for `RelWithDebInfo` which is just `-g` I think.
A LuaJIT debug build is not needed for this, is it?

> I did not measure the # of times get_node is called, I assume it is called for every node in the chunk. So very often.

That'd be bad because checking every node in the chunk is *the* usecase for vmanips.

> But it should not be lost that this simple change makes core.get_node() about 8x faster, without [...]

Sure, I didn't mean to say anything against optimizing that. `get_node` is a great target here because it's often used also outside of mapgen and in situations where bulk APIs aren't feasible to use.
Looking at the perf results I have no concrete complaints here.

**Comment by SmallJoker** [comment]
Why is it using `minetest.set_node` at all? For bulk node updates, VManip should be used for quicker operations thanks to lack of `content_id` to string resolution.

Related: I remember doing some tests using LuaJIT's FFI to reduce the double memory copy issue within (C++ -> Lua -> C++). pgimeno kindly provided a PR to implement the VManip more efficiently in #6863, including profiling. That showed possible speed-ups at around 2x (from LuaJIT to LuaJIT FFI).

**Comment by sfan5** [comment]
This idea has basically been implemented by b38ffdec279bcded98e34f5116c8d676aa9f73a7 and 310b12b5edaaa415e664a7766b9226fc95ce89e4.
I think this can be closed.
